x/params/upgrade_executor.go

```diff
@@ -0,0 +1,139 @@
+package params
+
+import (
+	"fmt"
+	"math"
+
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+	sdkerrors "github.com/okex/exchain/libs/cosmos-sdk/types/errors"
+	"github.com/okex/exchain/x/common"
+	govtypes "github.com/okex/exchain/x/gov/types"
+	"github.com/okex/exchain/x/params/types"
+)
+
+func NewUpgradeProposalHandler(k *Keeper) govtypes.Handler {
+	return func(ctx sdk.Context, proposal *govtypes.Proposal) sdk.Error {
+		switch c := proposal.Content.(type) {
+		case types.UpgradeProposal:
+			return handleUpgradeProposal(ctx, k, proposal.ProposalID, c)
+		default:
+			return common.ErrUnknownProposalType(DefaultCodespace, fmt.Sprintf("%T", c))
+		}
+	}
+}
+
+func handleUpgradeProposal(ctx sdk.Context, k *Keeper, proposalID uint64, proposal types.UpgradeProposal) sdk.Error {
+	curHeight := uint64(ctx.BlockHeight())
+	confirmHeight, err := getUpgradeProposalConfirmHeight(curHeight, proposal)
+	if err != nil {
+		return err
+	}
+	effectiveHeight := confirmHeight + 1
+
+	if curHeight < confirmHeight {
+		k.gk.InsertWaitingProposalQueue(ctx, confirmHeight, proposalID)
+		_ = storeWaitingUpgrade(ctx, k, proposal.UpgradeInfo, effectiveHeight) // ignore error
+		return nil
+	}
+
+	// proposal will be confirmed right now, check if ready.
+	cb, ready := k.queryReadyForUpgrade(proposal.Name)
+	if !ready {
+		// if no module claims that has ready for this upgrade,
+		// that probably means program's version is too low.
+		// To avoid status machine broken, we panic.
+		errMsg := fmt.Sprintf("there's a upgrade proposal named '%s' has been take effective, "+
+			"and the upgrade is incompatible, but your binary seems not ready for this upgrade. "+
+			"To avoid state machine broken, the program is panic. "+
+			"Using the latest version binary and re-run it to avoid this panic.", proposal.Name)
+		k.Logger(ctx).Error(errMsg)
+		panic(errMsg)
```

Comment:
Yes for the condition which called when executing a tx. 

For another condition, this function(`handleUpgradeProposal`) will also called from  `AppModule.EndBlock` of `gov` module to handle some waiting proposal that become effective in current block. When called from here the `panic` will work. And this may be the most common case.